text/0061-allocator-inside-of-runtime.md

```diff
@@ -0,0 +1,136 @@
+| --------------- | ------------------------------------------------------------------------------------------- |
+| **Start Date**  | 27 December 2023                                            |
+| **Description** | Supporting runtime built-in allocator makes the substrate runtime more versatile |
+| **Authors**     | Jiahao Ye |
+
+## Summary
+
+Currently, substrate runtime use an simple allocator defined by host side. Every runtime MUST
+import these allocator functions for normal execution. This situation make runtime code not versatile enough.
+
+So this RFC proposes to define a new spec for allocator part to make substrate runtime more generic.
+
+## Motivation
+
+Since this RFC define a new way for allocator, we now regard the old one as `legacy` allocator.
+As we all know, since the allocator implementation details are defined by the substrate client, parachain/parathread cannot customize memory allocator algorithm, so the new specification allows the runtime to customize memory allocation, and then export the allocator function according to the specification for the client side to use.
+Another benefit is that some new host functions can be designed without allocating memory on the client, which may have potential performance improvements. Also it will help provide a unified and clean specification if substrate runtime support multi-targets(e.g. RISC-V).
+There is also a potential benefit. Many programming languages that support compilation to wasm may not be friendly to supporting external allocator. This is beneficial for other programming languages ​​to enter the substrate runtime ecosystem.
+The last and most important benefit is that for offchain context execution, the runtime can fully support pure wasm. What this means here is that all imported host functions could not actually be called (as stub functions), then the various verification logic of the runtime can be converted into pure wasm, which provides the possibility for the substrate runtime to run block verification in other environments (such as in browsers and other non-substrate environments).
+
+## Stakeholders
+
+No attempt was made at convincing stakeholders.
+
+## Explanation
+
+### Runtime side spec
+
+This section contains a list of functions should be exported by substrate runtime.
+
+We define the spec as version 1, so the following `dummy` function `v1` MUST be exported to hint
+client that runtime is using version 1 spec, otherwise rollback to `legacy` allocator.
+The function should never be used, and its name is only for version checking.
+
+```wat
+  (export "v1" (func $v1))
+```
+
+Choose this way is more generic than custom section since many other tools do not support custom section very well. But if an environment want to run it, it should always be possible to parse
+the export section.
+
+The allocator functions are:
+
+```wat
+(export "alloc" (func $alloc))
+(export "dealloc" (func $dealloc))
+(export "realloc" (func $realloc))
+```
```

Comment:
We don't need these with: https://github.com/polkadot-fellows/RFCs/pull/4

Reply:
It's used by entrypoint call.